librosa/core/pitch.py

```diff
@@ -333,3 +337,456 @@ def piptrack(y=None, sr=22050, S=None, n_fft=2048, hop_length=None,
+    if win_length is None:
+        win_length = frame_length // 2
+
+    if win_length > frame_length:
+        raise ParameterError('win_length cannot exceed frame_length')
+
+    # Set the default hop if it is not already specified.
+    if hop_length is None:
+        hop_length = frame_length // 4
+
+    # Check that audio is valid.
+    util.valid_audio(y, mono=True)
+
+    # Pad the time series so that frames are centered
+    if center:
+        y = np.pad(y, frame_length // 2, mode=pad_mode)
+
+    # Frame audio.
+    y_frames = util.frame(y, frame_length=frame_length, hop_length=hop_length)
+
+    # Calculate minimum and maximum periods
+    min_period = max(int(np.floor(sr / fmax)), 2)
+    max_period = min(int(np.ceil(sr / fmin)), frame_length-win_length-1)
+
+    # Calculate cumulative mean normalized difference function.
+    yin_frames = _cumulative_mean_normalized_difference(
+        y_frames, frame_length, win_length, min_period, max_period)
+
+    # Parabolic interpolation.
+    parabolic_shifts = _parabolic_interpolation(yin_frames)
+
+    # Find Yin candidates and probabilities.
+    # The implementation here follows the official pYIN software which
+    # differs from the method described in the paper.
+    # 1. Define the prior over the thresholds.
+    thresholds = np.linspace(0, 1, n_thresholds + 1)
+    beta_cdf = scipy.stats.beta.cdf(
+        thresholds, beta_parameters[0], beta_parameters[1])
+    beta_probs = np.diff(beta_cdf)
+
+    yin_probs = np.zeros((yin_frames.shape[0]-2, yin_frames.shape[1]))
+    for i, yin_frame in enumerate(yin_frames.T):
+        # 2. For each frame find the troughs.
+        is_trough = util.localmax(-yin_frame, axis=0)
+        is_trough = is_trough[1:-1]
+        yin_frame = yin_frame[1:-1]
+        trough_index, = np.nonzero(is_trough)
+
+        if len(trough_index) == 0:
+            continue
+
+        # 3. Find the troughs below each threshold.
+        trough_heights = yin_frame[trough_index]
+        trough_thresholds = trough_heights[:, None] < thresholds[None, 1:]
+
+        # 4. Define the prior over the troughs.
+        # Smaller periods are weighted more.
+        trough_positions = np.cumsum(trough_thresholds, axis=0) - 1
+        n_troughs = np.count_nonzero(trough_thresholds, axis=0)
+        trough_prior = scipy.stats.boltzmann.pmf(
+            trough_positions, boltzmann_parameter, n_troughs)
+        trough_prior[~trough_thresholds] = 0
+
+        # 5. For each threshold add probability to global minimum if no trough is below threshold,
+        # else add probability to each trough below threshold biased by prior.
+        probs = np.sum(trough_prior * beta_probs, axis=1)
+        global_min = np.argmin(trough_heights)
+        n_thresholds_below_min = np.count_nonzero(~trough_thresholds[global_min, :])
+        probs[global_min] += no_trough_prob*np.sum(beta_probs[:n_thresholds_below_min])
+
+        yin_probs[trough_index, i] = probs
```

Comment:
I'm pretty sure this entire loop can be vectorized, but we can leave that for a separate PR.

The tricky part would be the boltzmann distribution, since it has a different `N` for each frame, but I think we could work around it.

Reply:
I did initially have it vectorized but it turned out to be slower and much worse on memory. Since there are only a few minima per frame I think you'd have to use sparse operations to get a speedup.

Reply:
Interesting, but good to know.  I guess the next route for optimization here would be to wrap it in numba to accelerate the loop.

The boltzmann pmf is still a barrier to that, but now because it's using scipy class code instead of pure numpy.  There's a new project to extend numba to scipy, but it appears to be a ways off.  We could replace this by reimplementing the [boltzmann pmf](https://github.com/scipy/scipy/blob/2ea67057b1bbce3a253af89a92ee93ffc9c97525/scipy/stats/_discrete_distns.py#L734-L738), since it's pretty simple.

I can take care of this as a separate PR for after we merge this one.  (Correctness first, then speed.)